### ISSUE TITLE
build failures on docs.rs

### ISSUE THREAD

**Issue by ckaran** [emilk/egui#146]
You're getting build failures on docs.rs for eframe (and maybe others?)  Here is a link to the log file:

https://docs.rs/crate/eframe/0.8.0/builds/338677

I didn't see anything else that jumped out at me, but I'm not as familiar with the project as I wish I was.

**Comment by emilk** [comment]
Thanks for noticing! It seems `egui_glium` docs have been broken from the very start. Here is 0.1.0 from august 2020: https://docs.rs/crate/egui_glium/0.1.3/builds/279591

The build error is rather cryptic. I'll put it down as yet another reason to switch from glium (https://github.com/emilk/egui/issues/93)

**Comment by follower** [comment]
[WIP: Plan to add some more details.]

From a brief investigation[1] I noticed:

 * There is a more recent version of the `xcb` crate (the crate that seems to be cause of the issue). (`0.9.0` vs `0.8.2`)

 * The `docs.rs` help (linked in the orange build failure notice on the crate page) has a section "[Write attempt on read-only directories](https://docs.rs/about/builds#read-only-directories)" that mentions:

    > _Most of the sandbox is a read-only file system, including the source directory of your crate and its dependencies. If your build.rs generates files that are relevant for documentation, consider writing to the _cargo output directory_, passed in the environment variable `OUT_DIR`._ 

 * The same page also mentions in <https://docs.rs/about/builds#detecting-docsrs>:

    >  _To recognize Docs.rs from build.rs files, you can test for the environment variable `DOCS_RS`, [...]  This approach can be helpful if you need dependencies for building the library, but not for building the documentation._  

So, it seems two options for solving/working around could be:

 1. Bump version of `xcb` required to latest.
 2. Detect when building docs only and/or on `DOC_RS` and skip the build step/dependency (assuming it's not required for doc generation.)

----

[1] Key sections of the logs are:

* <https://docs.rs/crate/eframe/0.8.0/builds/338677>:

```
[INFO] [stderr] error: failed to run custom build command for `xcb v0.8.2`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/xcb-4495dea23494d888/build-script-build` (exit code: 101)
[INFO] [stderr]   --- stderr
[INFO] [stderr]   error occured in handler:  ('request: ', ('xcb', 'SELinux', 'GetClientContext'))
[INFO] [stderr]   Traceback (most recent call last):
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 2448, in <module>
[INFO] [stderr]       module.generate()
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/xcbgen/state.py", line 108, in generate
[INFO] [stderr]       self.close()
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 296, in rs_close
[INFO] [stderr]       _f.writeout(os.path.join(module.rs_srcdir, "ffi", "%s.rs" % _module_name(_ns.ext_name)))
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 114, in writeout
[INFO] [stderr]       with open(path, 'w') as f:
[INFO] [stderr]   OSError: [Errno 30] Read-only file system: '/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/src/ffi/selinux.rs'
[INFO] [stderr]   Error in sys.excepthook:
[INFO] [stderr]   Traceback (most recent call last):
[INFO] [stderr]     File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 119, in apport_excepthook
[INFO] [stderr]       pr.add_user_info()
[INFO] [stderr]     File "/usr/lib/python3/dist-packages/apport/report.py", line 424, in add_user_info
[INFO] [stderr]       user = pwd.getpwuid(os.getuid())[0]
[INFO] [stderr]   KeyError: 'getpwuid(): uid not found: 1001'
[INFO] [stderr] 
[INFO] [stderr]   Original exception was:
[INFO] [stderr]   Traceback (most recent call last):
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 2448, in <module>
[INFO] [stderr]       module.generate()
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/xcbgen/state.py", line 108, in generate
[INFO] [stderr]       self.close()
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 296, in rs_close
[INFO] [stderr]       _f.writeout(os.path.join(module.rs_srcdir, "ffi", "%s.rs" % _module_name(_ns.ext_name)))
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 114, in writeout
[INFO] [stderr]       with open(path, 'w') as f:
[INFO] [stderr]   OSError: [Errno 30] Read-only file system: '/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/src/ffi/selinux.rs'
[INFO] [stderr]   thread 'main' panicked at 'processing of /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/xml/xselinux.xml returned non-zero (1)', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/build.rs:75:17
[INFO] [stderr]   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

* <https://docs.rs/crate/egui_glium/0.1.3/builds/279591>:
```
[INFO] [stderr] error: failed to run custom build command for `xcb v0.8.2`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/xcb-e7147420ca72501a/build-script-build` (exit code: 101)
[INFO] [stderr]   --- stderr
[INFO] [stderr]   error occured in handler:  ('request: ', ('xcb', 'SELinux', 'GetClientContext'))
[INFO] [stderr]   Traceback (most recent call last):
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 2448, in <module>
[INFO] [stderr]       module.generate()
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/xcbgen/state.py", line 108, in generate
[INFO] [stderr]       self.close()
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 296, in rs_close
[INFO] [stderr]       _f.writeout(os.path.join(module.rs_srcdir, "ffi", "%s.rs" % _module_name(_ns.ext_name)))
[INFO] [stderr]     File "/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/rs_client.py", line 114, in writeout
[INFO] [stderr]       with open(path, 'w') as f:
[INFO] [stderr]   OSError: [Errno 30] Read-only file system: '/opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/src/ffi/selinux.rs'
[INFO] [stderr]   thread 'main' panicked at 'processing of /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/xml/xselinux.xml returned non-zero (1)', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/xcb-0.8.2/build.rs:75:17
[INFO] [stderr]   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by follower** [comment]
However, in terms of reasons to switch, I did notice that it appears there are some other, security related concerns with `xcb` (but it may still be a dependency of alternatives, also):

 * https://github.com/rtbo/rust-xcb/issues/96

**Comment by follower** [comment]
**TL;DR: Workaround is probably to override `rust-clipboard` dependency on `x11-clipboard` to version `0.5.1`.**

### Details

`cargo tree` reveals:

```
│   ├── egui_glium v0.8.0
│   │   ├── clipboard v0.5.0
│   │   │   └── x11-clipboard v0.3.3
│   │   │       └── xcb v0.8.2
│   │   │           ├── libc v0.2.80
│   │   │           └── log v0.4.11
│   │   │               └── cfg-if v0.1.10
│   │   │           [build-dependencies]
│   │   │           └── libc v0.2.80
│   │   ├── egui v0.8.0 (*)
```

Which lead to discovering:

 * "[Update x11-clipboard dependency to use xcb [0.9.0] ~~0.5.1~~, fixes docs.rs builds #76](https://github.com/aweinstock314/rust-clipboard/pull/76)"
  
   Which is an _unmerged PR from a year ago_ bumping `rust-clipboard`'s `x11-clipboard` dependency version to [`0.5.1`](https://crates.io/crates/x11-clipboard/0.5.1) which depends on `xcb` `0.9.0` which is seemingly fixed.

**Comment by MatchaChoco010** [comment]
For your information, copypasta is a fork in the rust-clipboard crate that is maintained.
You can find it at https://github.com/alacritty/copypasta
My egui integration changed to use copypasta from rust-clipboard and my docs.rs now builds correctly.

**Comment by follower** [comment]
Nice, thanks for mentioning copypasta[0]!

Can confirm the version bump is here:

 * https://github.com/alacritty/copypasta/blob/f4125b09c30bee525327660e27411b76ddbe0718/Cargo.toml#L28

The README mentions also adding support for Wayland clipboard so not sure if that would cause any unexpected issues elsewhere.

Good to know that the docs now at least build for you.

----

[0] I had _just_ noticed the `copypasta` dependency name earlier today while looking at some crates but hadn't actually looked to find out what the crate did! So it's amusing that it turned out to be a fix for this unrelated issue. :)
